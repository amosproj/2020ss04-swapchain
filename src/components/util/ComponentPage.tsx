import React from "react"

interface Props {
  title?: string
  children: any
  bg?: string
}

export const ComponentPage = (props: Props) => {
  return (
    <div className="container mx-auto">
      <main className={`pb-10 mx-auto ${props.bg} ${props.bg ? "shadow-sm" : ""}  p-4 sm:px-6 lg:px-8`}>
        <h1 className="py-4 text-3xl font-semibold leading-tight text-gray-900">{props.title}</h1>
        {props.children}
      </main>
    </div>
  )
}
