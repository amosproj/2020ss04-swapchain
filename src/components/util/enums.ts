export enum State {
  IDLE,
  RUNNING,
  SUCCESS,
  ERROR,
  FAILURE,
}

export enum Currency {
  BTC = "BTC",
  BTS = "BTS",
}

export enum Network {
  MAINNET = "mainnet",
  TESTNET = "testnet",
}

export enum Priority {
  HIGH,
  MEDIUM,
  LOW,
}
