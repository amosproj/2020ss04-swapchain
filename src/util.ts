import * as bitcoin from "bitcoinjs-lib"

export const fakeKey = (length: number, network?: string) => {
  let prefix = ""
  if (network === "mainnet") {
    prefix = "L"
  } else if (network === "testnet") {
    prefix = "c"
  }

  return prefix + hash(Math.random().toString()).substring(0, length)
}

export const hash = (s: string): string => {
  return bitcoin.crypto.hash256(Buffer.from(s)).toString("hex")
}

export const toPublicKey = (privateKey: string): string => {
  const keyPair = bitcoin.ECPair.fromWIF(privateKey, bitcoin.networks.testnet)
  const keyPairCompressed = bitcoin.ECPair.fromPrivateKey(keyPair.privateKey!, { compressed: true })
  return keyPairCompressed.publicKey.toString("hex")
}
