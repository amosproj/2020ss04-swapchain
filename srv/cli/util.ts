import readline from "readline"

/**
 * Get input from stdin
 *
 * @param str - A question to ask the user.
 * @returns answer from the user
 */
export async function askUser(str: string): Promise<string> {
  const read = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  })

  return new Promise((resolve) => read.question("\u001b[36;1m" + str + "\u001b[0m", resolve))
}
