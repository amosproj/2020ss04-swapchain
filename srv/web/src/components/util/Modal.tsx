import React from "react"
import { ReactComponent as X } from "../../icons/x.svg"
type Props = {
    close: () => void
    open: boolean
    children: any
    title: string
}


export default (props: Props) => {
  
    return (
        <div className={props.open ? "" : "hidden"}>
            <div className="absolute inset-0 bg-gray-500 opacity-75"></div>
            <div className="absolute inset-0 z-10 flex items-center justify-center">
                <div className="p-12 bg-white rounded-sm">
                    <header className="flex items-center justify-between">
                    <h1 className="text-lg font-bold text-gray-500">{props.title}</h1>
                        <X className="w-8 h-8 text-gray-400 duration-200 ease-in-out hover:text-gray-800" onClick={props.close}></X>
                    </header>
                    <div className="flex flex-col items-center justify-around md:flex-row ">
                    {props.children}
                </div>

            </div>
        </div>
        </div>
    )
}

