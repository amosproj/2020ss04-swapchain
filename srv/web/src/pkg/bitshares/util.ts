import { Apis as btsWebsocketApi, ChainConfig } from "bitsharesjs-ws"
import { PrivateKey } from "bitsharesjs"

export async function getAccount(privateKey: string, node: string, network: string): Promise<string> {

    if (privateKey.length !== 51) {
        throw new Error("Invalid private key length.")
    }

    // Set chain IDs to get correct prefix for public key
    if (network === "mainnet") {
        ChainConfig.setChainId("4018d7844c78f6a6c41c6a552b898022310fc5dec06da467ee7905a8dad512c8")
    } else if (network === "testnet") {
        ChainConfig.setChainId("39f5e2ede1f8bc1a3a54a7914414e3779e33193f1f5693510e73cb7a87617447")
    } else {
        throw new Error("Invalid network name. Choose mainnet or testnet.")
    }

    const publicKey = PrivateKey.fromWif(privateKey).toPublicKey().toString()

    await btsWebsocketApi.instance(node, true).init_promise

    const accounts = await btsWebsocketApi.db.get_key_references([publicKey])

    if (typeof accounts[0][0] === "undefined") {
        throw new Error(`Could not find any accounts for private key ${privateKey} on ${network}`)
    }

    const accountID = accounts[0][0]

    const accountNames = await btsWebsocketApi.db.get_accounts([accountID])

    if (Object.keys(accountNames[0]).length === 0) {
        throw new Error(`Could not get accounts for account ID ${accountID} on ${network}`)
    }
    
    return accountNames[0].name
}