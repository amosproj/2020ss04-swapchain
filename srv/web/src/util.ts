import crypto from "crypto"

export const fakeKey = (length: number, network: string) => {
  let prefix: string
  if (network === "mainnet") {
    prefix = "L"
  } else if (network === "testnet") {
    prefix = "c"
  } else {
    throw new Error("Invalid network name. Choose mainnet or testnet.")
  }
  
  return prefix + hash(Math.random().toString()).substring(0, length) + "..."
}

export const hash = (s: string): string => {
  return crypto.createHash("SHA256").update(s).digest("hex")
}
