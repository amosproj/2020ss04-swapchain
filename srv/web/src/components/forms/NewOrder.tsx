import React, { useState, useEffect } from "react"
import { fakeAddress } from "../../util"
import { CryptoChoice } from "./CryptoChoice"
import { PriorityChoice } from "./PriorityChoice"
import { Label } from "./Label"
import { Input } from "./Input"
import { SubmitButton } from "./SubmitButton"
export const NewOrder = () => {
    enum Currency {
        BTC,
        BTS,
    }

    enum Priority {
        HIGH,
        MEDIUM,
        LOW,
    }

    const rateUnit = (): string[] => {
        switch (fields.currencyToGive) {
            case Currency.BTC:
                return ["BTS", "BTC"]

            case Currency.BTS:
                return ["BTC", "BTS"]
        }
    }

    const [fields, setFields] = useState({
        currencyToGive: Currency.BTC,
        amountToSend: 0,
        rate: 0,
        amountYouReceive: 0,
        bitcoinPublicAddress: "",
        bitsharesAccountName: "",
        timelockDuration: 6,
        orderDuration: 7,
        priority: Priority.HIGH,
    })
    const [isValid, setValid] = useState(false)
    const [showError, setShowError] = useState(false)


    // calculate received fudns
    useEffect(() => {
        setFields({
            ...fields,
            amountYouReceive: fields.amountToSend * fields.rate
        })
    }, [fields.amountToSend, fields.rate])

    //validation
    useEffect(() => {
        const isValid = fields.amountToSend > 0 &&
            fields.rate > 0 &&
            fields.bitcoinPublicAddress !== "" &&
            fields.bitsharesAccountName !== "" &&
            fields.timelockDuration > 0 &&
            fields.orderDuration > 0 &&
            fields.amountYouReceive > 0

        setValid(isValid)
        if (isValid) {
            setShowError(false)
        }
    }, [fields])

    const updateField = (e: React.FormEvent<HTMLInputElement>) => {
        setFields({
            ...fields,
            [e.currentTarget.name]: e.currentTarget.value
        })
    }
    const updateFieldByName = (name: string, value: number | string) => {
        setFields({
            ...fields,
            [name]: value
        })
    }

    const submitHandler = () => {
        if (!isValid) {
            setShowError(true)
            return
        }
        console.log(fields)
    }

    return (
        <div className="container mx-auto">
            <section className="mt-8">
                <Label label="What do you want to trade"></Label>
                <div className="flex items-center -mx-3">

                    <button className="flex-grow px-3 focus:outline-none" onClick={() => updateFieldByName("currencyToGive", Currency.BTC)}>
                        <CryptoChoice label="BTC" name="Bitcoin" selected={fields.currencyToGive === Currency.BTC}></CryptoChoice>
                    </button>
                    <button className="flex-grow px-3 focus:outline-none" onClick={() => updateFieldByName("currencyToGive", Currency.BTS)}>
                        <CryptoChoice label="BTS" name="Bitshares" selected={fields.currencyToGive === Currency.BTS}></CryptoChoice>
                    </button>
                </div>
            </section>

            <section className="items-center justify-between mt-12 -mx-3 md:flex">

                <div className="flex-grow p-2">
                    <Label label="Amount to send"></Label>
                    <div className="relative">
                        <span className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600">
                            {rateUnit()[0]}           </span>
                        <Input name="amountToSend" value={fields.amountToSend} onChange={updateField} type="number" placeholder="0.00000"></Input>
                    </div>
                </div>
                <div className="flex-grow p-2">
                    <Label label="Rate"></Label>
                    <div className="relative">
                        <div className="absolute inset-y-0 right-0 flex flex-col items-center justify-center mr-6 text-xs text-gray-600">
                            <span>{rateUnit()[1]}</span>
                            <span className="border-t border-gray-500">{rateUnit()[0]}</span>
                        </div>
                        <Input name="rate" value={fields.rate} onChange={updateField} type="number" placeholder="0.00000"></Input>
                    </div> </div>
                <div className="flex-grow p-2">
                    <Label label="excluding fees you will receive"></Label>
                    <div className="relative">
                        <span className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600">
                            {rateUnit()[1]}           </span>
                        <span className="block w-full py-3 text-center text-gray-700 rounded">{fields.amountYouReceive}</span>
                    </div>
                </div>
            </section>

            <section className="mt-12 -mx-3 md:flex">

                <div className="px-3 md:w-1/2">
                    <Label label="Bitcoin public address"></Label>
                    <Input name="bitcoinPublicAddress" value={fields.bitcoinPublicAddress} onChange={updateField} type="text" placeholder={fakeAddress(30)}></Input>
                </div>
                <div className="px-3 mt-4 md:w-1/2 md:mt-0">
                    <Label label="Bitshares account name"></Label>
                    <Input name="bitsharesAccountName" value={fields.bitsharesAccountName} onChange={updateField} type="text" placeholder="amos"></Input>
                </div>
            </section>

            <section className="mt-24 -mx-3 md:flex">

                <div className="px-3 md:w-1/2">
                    <Label label="Timelock duration (1 block is roughly 10 minutes)"></Label>
                    <div className="relative">
                        <span className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600">
                            Blocks          </span>
                        <Input name="timelockDuration" value={fields.timelockDuration} onChange={updateField} type="number" placeholder="6"></Input>
                    </div>
                </div>
                <div className="px-3 mt-4 md:w-1/2 md:mt-0">
                    <Label label="Order duration"></Label>
                    <div className="relative">
                        <span className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600">
                            Days</span>
                        <Input name="orderDuration" value={fields.orderDuration} onChange={updateField} type="number" placeholder="how long your order will stay active"></Input>
                    </div>
                </div>
            </section>
            <section className="mt-12">
                <Label label="Choose your priority"></Label>
                <div className="flex items-center -mx-3">
                    <button className="w-1/3 px-3 focus:outline-none" onClick={() => updateFieldByName("priority", Priority.HIGH)}>
                        <PriorityChoice label={Priority[Priority.HIGH]} description="You pay the highest fees to increase the chance for your transaction to be picked up by the miners." selected={fields.priority === Priority.HIGH}></PriorityChoice>
                    </button>
                    <button className="w-1/3 px-3 focus:outline-none" onClick={() => updateFieldByName("fields.", Priority.MEDIUM)}>
                        <PriorityChoice label={Priority[Priority.MEDIUM]} description="You pay a moderate amount of fees so miners will probably confirm your transaction soon." selected={fields.priority === Priority.MEDIUM}></PriorityChoice>
                    </button>
                    <button className="w-1/3 px-3 focus:outline-none" onClick={() => updateFieldByName("priority", Priority.LOW)}>
                        <PriorityChoice label={Priority[Priority.LOW]} description="You pay the lowest fees but might have to wait a few more blocks for your transaction to be confirmed." selected={fields.priority === Priority.LOW}></PriorityChoice>
                    </button>
                </div>
            </section>

            <footer className="pt-12 mt-8 text-center border-t border-gray-300 f">
                <p className={`-mt-4 pb-4 text-red-500 text-sm ${showError ? "" : "hidden"}`}>Please fill in all fields</p>
                <SubmitButton color={isValid ? "teal" : "gray"} label="Submit" onClick={submitHandler}></SubmitButton>


            </footer>
        </div>
    )
}