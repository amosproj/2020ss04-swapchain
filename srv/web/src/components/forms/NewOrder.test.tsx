import React from 'react';
import {NewOrder} from './NewOrder';
import renderer from 'react-test-renderer';
import * as util from "../../util"
jest.mock("../../util")

jest.spyOn(util, "fakeAddress").mockReturnValue("randomLookingHash")
it('renders correctly', () => {
  const tree = renderer
    .create(<NewOrder></NewOrder>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});