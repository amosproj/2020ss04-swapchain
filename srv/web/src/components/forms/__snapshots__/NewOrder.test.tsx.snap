// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container mx-auto"
>
  <section
    className="mt-8"
  >
    <label
      className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
    >
      What do you want to trade
    </label>
    <div
      className="flex items-center -mx-3"
    >
      <button
        className="flex-grow px-3 focus:outline-none"
        onClick={[Function]}
      >
        <div
          className="relative flex justify-between items-center border rounded border-teal-400 shadow"
        >
          <div
            className="flex flex-col items-start p-4"
          >
            <span
              className="text-left font-bold text-gray-900"
            >
              Bitcoin
            </span>
            <p
              className="text-sm text-left text-gray-600"
            >
              You are giving Bitcoin away
            </p>
          </div>
          <div
            className="flex flex-col items-center justify-center"
          >
            <span
              className="p-1 m-1 text-xs font-thin text-gray-500"
            />
            <span
              className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-900 "
            >
              BTC
            </span>
          </div>
        </div>
      </button>
      <button
        className="flex-grow px-3 focus:outline-none"
        onClick={[Function]}
      >
        <div
          className="relative flex justify-between items-center border rounded border-gray-200 bg-gray-100"
        >
          <div
            className="flex flex-col items-start p-4"
          >
            <span
              className="text-left font-bold text-gray-600"
            >
              Bitshares
            </span>
            <p
              className="text-sm text-left text-gray-400"
            />
          </div>
          <div
            className="flex flex-col items-center justify-center"
          >
            <span
              className="p-1 m-1 text-xs font-thin text-gray-500"
            />
            <span
              className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-300"
            >
              BTS
            </span>
          </div>
        </div>
      </button>
    </div>
  </section>
  <section
    className="items-center justify-between mt-12 md:space-x-4 md:flex"
  >
    <div
      className="flex-grow"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Amount to send
      </label>
      <div
        className="relative"
      >
        <span
          className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600"
        >
          BTC
                     
        </span>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          min={0}
          name="amountToSend"
          onChange={[Function]}
          placeholder="0.00000000"
          required={true}
          step={1e-8}
          type="number"
          value=""
        />
      </div>
    </div>
    <div
      className="flex-grow"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Rate
      </label>
      <div
        className="relative"
      >
        <div
          className="absolute inset-y-0 right-0 flex flex-col items-center justify-center mr-6 text-xs text-gray-600"
        >
          <span>
            BTS
          </span>
          <span
            className="border-t border-gray-500"
          >
            BTC
          </span>
        </div>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          min={0}
          name="rate"
          onChange={[Function]}
          placeholder="0.0000"
          required={true}
          step={1e-8}
          type="number"
          value=""
        />
      </div>
       
    </div>
    <div
      className="flex-grow"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        excluding fees you will receive
      </label>
      <div
        className="relative"
      >
        <span
          className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600"
        >
          BTS
                     
        </span>
        <span
          className="block w-full py-3 text-center text-gray-700 border border-gray-200 rounded focus:border-teal-500"
        >
          0
        </span>
      </div>
    </div>
  </section>
  <section
    className="mt-12 md:space-x-4 md:flex"
  >
    <div
      className=" md:w-1/2"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Bitcoin public address
      </label>
      <input
        className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
        name="bitcoinPublicAddress"
        onChange={[Function]}
        placeholder="randomLookingHash"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="mt-4 md:w-1/2 md:mt-0"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Bitshares account name
      </label>
      <input
        className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
        name="bitsharesAccountName"
        onChange={[Function]}
        placeholder="amos"
        required={true}
        type="text"
        value=""
      />
    </div>
  </section>
  <section
    className="mt-24 md:space-x-4 md:flex"
  >
    <div
      className=" md:w-1/2"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Timelock duration (1 block is roughly 10 minutes)
      </label>
      <div
        className="relative"
      >
        <span
          className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600"
        >
          Blocks          
        </span>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          max={20}
          min={6}
          name="timelockDuration"
          onChange={[Function]}
          placeholder="6"
          required={true}
          type="number"
          value={6}
        />
      </div>
    </div>
    <div
      className="mt-4 md:w-1/2 md:mt-0"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Order duration
      </label>
      <div
        className="relative"
      >
        <span
          className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600"
        >
          Days
        </span>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          max={7}
          min={1}
          name="orderDuration"
          onChange={[Function]}
          placeholder="how long your order will stay active"
          required={true}
          type="number"
          value={7}
        />
      </div>
    </div>
  </section>
  <section
    className="mt-12"
  >
    <label
      className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
    >
      Choose your priority
    </label>
    <div
      className="flex items-center -mx-3"
    >
      <button
        className="w-1/3 px-3 focus:outline-none"
        onClick={[Function]}
      >
        <div
          className="relative flex justify-between items-center border rounded border-teal-400 shadow"
        >
          <div
            className="flex flex-col items-start p-4"
          >
            <span
              className="text-left font-bold text-gray-900"
            >
              HIGH
            </span>
            <p
              className="text-sm text-left text-gray-600"
            >
              You pay the highest fees to increase the chance for your transaction to be picked up by the miners.
            </p>
          </div>
          <div
            className="flex flex-col items-center justify-center"
          >
            <span
              className="p-1 m-1 text-xs font-thin text-gray-500"
            />
            <span
              className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-900 "
            />
          </div>
        </div>
      </button>
      <button
        className="w-1/3 px-3 focus:outline-none"
        onClick={[Function]}
      >
        <div
          className="relative flex justify-between items-center border rounded border-gray-200 bg-gray-100"
        >
          <div
            className="flex flex-col items-start p-4"
          >
            <span
              className="text-left font-bold text-gray-600"
            >
              MEDIUM
            </span>
            <p
              className="text-sm text-left text-gray-400"
            >
              You pay a moderate amount of fees so miners will probably confirm your transaction soon.
            </p>
          </div>
          <div
            className="flex flex-col items-center justify-center"
          >
            <span
              className="p-1 m-1 text-xs font-thin text-gray-500"
            />
            <span
              className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-300"
            />
          </div>
        </div>
      </button>
      <button
        className="w-1/3 px-3 focus:outline-none"
        onClick={[Function]}
      >
        <div
          className="relative flex justify-between items-center border rounded border-gray-200 bg-gray-100"
        >
          <div
            className="flex flex-col items-start p-4"
          >
            <span
              className="text-left font-bold text-gray-600"
            >
              LOW
            </span>
            <p
              className="text-sm text-left text-gray-400"
            >
              You pay the lowest fees but might have to wait a few more blocks for your transaction to be confirmed.
            </p>
          </div>
          <div
            className="flex flex-col items-center justify-center"
          >
            <span
              className="p-1 m-1 text-xs font-thin text-gray-500"
            />
            <span
              className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-300"
            />
          </div>
        </div>
      </button>
    </div>
  </section>
  <footer
    className="pt-12 mt-8 text-center border-t border-gray-300 f"
  >
    <p
      className="-mt-4 pb-4 text-red-500 text-sm hidden"
    >
      Please fill in all fields
    </p>
    <button
      className="bg-gray-100 border-gray-400 inline-flex px-8 py-3 font-semibold text-gray-900 uppercase transition duration-200 ease-in-out border focus:outline-none rounded-sm hover:bg-white"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </footer>
</div>
`;
